extract dispatch keys from optional Tensors (unboxed)

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58296 extract dispatch keys from optional Tensors (unboxed)**

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58296 extract dispatch keys from optional Tensors (unboxed)**

Correctly extract dispatch keys from optional Tensors in unboxed calls. Companion to #58283 which does the same for boxed calls.

E.g. fixes the following repro case (from @zou3519 and @Chillee):

```
import torch

# make a quantized tensor
r = torch.rand(3, 2, dtype=torch.float) * 4 - 2
scale = 0.02
zero_point = 2
quantized = torch.quantize_per_tensor(r, scale, zero_point, torch.quint8)

# The following invokes the quantized key's version of the clamp.
# The quantized key doesn't actually have a kernel registered for clamp,
# so it throws an error to us.
torch.clamp(quantized, quantized, quantized)

# However, the following straight up calls into native::clamp (e.g. the CPU
# kernel) instead of dispatching on the quantized key and giving us an error:
cpu_tensor = torch.randn(3, 2)
torch.clamp(cpu_tensor, quantized, quantized)
```


Differential Revision: [D28436822](https://our.internmc.facebook.com/intern/diff/D28436822)

Commit:
extract dispatch keys from optional Tensors (unboxed)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28436822).

Chillee commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -55,6 +55,11 @@ namespace detail {
       ts = ts | x.key_set();
     }
+    void operator()(c10::optional<at::Tensor> x) {
+      if (x.has_value() && x->defined()) {
```
Why does this need to check if it's defined but the above one doesn't?

ezyang replied:
+1 @Chillee  here, the defined test is unnecessary, because (1) you aren't supposed to pass undefined tensors in and, (2) even if you pass it in, undefined tensors have defined key set which is just empty bitset.

bhosmer replied:
yup agree, will remove

Chillee reviewed (approved):
This fix solves my issues :)

Commit:
Update on "extract dispatch keys from optional Tensors (unboxed)"



Correctly extract dispatch keys from optional Tensors in unboxed calls. Companion to #58283 which does the same for boxed calls.

E.g. fixes the following repro case (from @zou3519 and @Chillee):

```
import torch

# make a quantized tensor
r = torch.rand(3, 2, dtype=torch.float) * 4 - 2
scale = 0.02
zero_point = 2
quantized = torch.quantize_per_tensor(r, scale, zero_point, torch.quint8)

# The following invokes the quantized key's version of the clamp.
# The quantized key doesn't actually have a kernel registered for clamp,
# so it throws an error to us.
torch.clamp(quantized, quantized, quantized)

# However, the following straight up calls into native::clamp (e.g. the CPU
# kernel) instead of dispatching on the quantized key and giving us an error:
cpu_tensor = torch.randn(3, 2)
torch.clamp(cpu_tensor, quantized, quantized)
```


Differential Revision: [D28436822](https://our.internmc.facebook.com/intern/diff/D28436822)

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28436822).
